apps/docs/pages/guides/auth/auth-postgres-hooks.mdx

```diff
@@ -0,0 +1,669 @@
+
+
+Below is a flow chart illustrating how Supabase Auth makes use of an Auth Hook:
+
+<Mermaid
+ chart={`flowchart TD
+    subgraph Supabase Auth
+    Start[Event Occurs] --> DecisionA{Is a Hook defined?}
+
+    ValidityCheck --Yes-->ContinueFlow[Continue with flow]
+    ValidityCheck --No-->Error(Return Error)
+
+    DecisionA -- No ---->H[Continue with Supabase Default Behaviour]
+    H--> ValidityCheck{Is response valid?}
+    DecisionA --Yes--->ValidateInput((Validate Inputs))
+        ValidateOutput -->ContinueFlow
+    end
+
+
+
+    subgraph Auth Hook
+    ValidateInput ---->CallHook[Execute PostgreSQL Function]
+
+    CallHook ---> ValidateOutput((Validate Outputs))
+    end`}
+    />
+
+
+
+## Use Cases
+
+Based on user feedback, we decided on a few use cases to illustrate how a hook works. These use cases are not exhaustive and you can opt to [write your own hook](#writing-your-own-hook) if you have a use case which doesn’t fit any of the examples below.
+
+## MFA Verification Attempt Hook
+
+Depending on your security or compliance requirements, you may need to enforce stricter restrictions beyond the default [Supabase MFA implementation](https://supabase.com/docs/guides/auth/auth-mfa).
+
+For such scenarios, the MFA Verification Attempt hook can allow you to add additional checks to:
+
+- Ask users to wait a set time between verification attempts, to limit the number of attempts that can be done in a period of time.
+- Force sign out users if they have too many invalid verification attempts.
+- Count, rate limit, ban or do other actions.
+
+### Inputs
+
+Supabase Auth will supply the following fields to your hook:
+
+- `factor_id` Unique identifier for the MFA factor being verified. 
+- `user_id` Unique identifier for the user.
+- `valid` Whether the verification attempt was valid. For TOTP, this means that the six digit code was correct (true) or incorrect (false). The hook will generally receive true once, and false many times for the same user and factor.
+
+Example payload:
+
+```json
+{
+	"factor_id": "6eab6a69-7766-48bf-95d8-bd8f606894db",
+	"user_id": "3919cb6e-4215-4478-a960-6d3454326cec",
+	"valid": true,
+}
+```
+
+### Outputs
+
+In exchange your hook should return a JSON object with these fields set:
+
+********************************On Success********************************
+
+Return these only if your hook processed the input without errors.
+
+- `decision` A string containing the decision on whether to allow authentication to move forward. Use `reject` to completely reject the verification attempt and log the user out of all active sessions. Use `continue` to use the default Supabase Auth behavior.
+- `message` The message to show the user if the decision was `reject`.
+
+Example output:
+
+```json
+{
+   "decision": "reject",
+   "message" : "You have exceeded maximum number of MFA attempts."
+}
+```
+
+****************On Error****************
+
+Return this if your hook processed the input with some error.
```

Comment:
I'm not entirely clear on what should be a rejection versus an error. I thought perhaps an error would be a server error (application's fault) while a rejection is a rule violation (user's fault), but this example of verifying a factor too often sounds more like a user error, in line with the too many invalid attempts above